### DISCUSSION TITLE
Question: is there much difference/lacking features of the Serilog.Sentry library (vs main Sentry SDK)?

### DISCUSSION THREAD

**Discussion by PureKrome** [getsentry/sentry-dotnet#894] [Q&A]
G'Day 👋🏻 !

With the `Sentry.Serilog` library, is there much features missing .. versus using the main `Sentry` library?

Context: I prefer to pass `ILogger<T>` all around my application. In my `ConfigureServices`, I would prefer to just setup Serilog and then log whatever to wherever, via sinks, like `Sentry.Serilog`. 

So if I get an error, I would then log that to Sentry.

Is it safe to assume that the data which is logged to sentry via the Serilog sink is pretty much feature rich as manually using the Serilog library? Like, rich exception information, etc?

Or in other words, if I use the `Sentry.Serilog` library, I would be missing out of the following pieces of data:
- `<something>`
- `<another thing>`
- etc.

**Comment by bruno-garcia** [comment]
For the main `Sentry` package, you're not missing anything.

Except that you can use the Sentry API to add more data to the event. Like `Contexts` or `tags` which get indexed.

The protocol has well defined stuff: https://develop.sentry.dev/sdk/event-payloads/
C# API Docs: https://getsentry.github.io/sentry-dotnet/api/Sentry.Protocol.html

But the integrations such as `UnhandledException` and `UnobservedTaskException` are hooked up automatically so even if you're using `Sentry.Serilog`, such errors would be captured.

For more context, check out the [Sentry Serilog Sink](https://github.com/getsentry/sentry-dotnet/blob/fdf24546995fd0a1e9c009668647c8648af9e3ec/src/Sentry.Serilog/SentrySink.cs#L58-L136), and you'll see what we **do use**. And if you check out the API docs, and the [Sentry Docs](https://docs.sentry.io/platforms/dotnet/enriching-events/context/) you'll what possible through the SDK itself (not limited by a logging integration).

If you're on a ASP.NET Core app, and don't install the `Sentry.AspNetCore` you'll be missing:

* Request information like headers
* User agent is used by Sentry to show browser, OS info
* Errors handled by the error page
* Exceptions thrown in the middleware pipeline
* User authenticated in ASP.NET Core is 
* .. other that I can't think of right now

Similar data for ASP.NET classic (which btw, is moving to a `Sentry.AspNet` dedicated package on Sentry SDK 3.0.0.

**Comment by PureKrome** [comment]
Thanks @bruno-garcia for the help/reply.

So if i understand you right ...

- `Sentry` nuget and `Sentry.Serilog` nuget offer the same functionality. Of course, the Serilog one is just a wrapper.

[Quote from docs](https://docs.sentry.io/platforms/dotnet/guides/serilog/):

> This package extends Sentry main SDK. That means that besides the logging related features, through this package you'll also get access to all API and features available in the main Sentry SDK.

kewl!

> Except that you can use the Sentry API to add more data to the event. 

Does this mean that if you just use the `Sentry` nuget, you can do more stuff (add a context and tags) while the `Serilog.Sentry` nuget hasn't exposed this?

> If you're on a ASP.NET Core app, and don't install the Sentry.AspNetCore you'll be missing:

So I like using Serilog because it has the concept of sinks. So (for example) I can get logs going to 
- console
- seq
- sentry 

and only have one MS `ILogger` passed around to my code and various reusable nuget libraries that are logging agnostic.

So this means I would be missing out on these lovely extra's, for an ASP.NET Core app, because I would only be installed `Serilog.Sentry`. Is there any reason why there isn't a `Serilog.Sentry.AspNetCore` nuget?

**Comment by bruno-garcia** [comment]
> > Except that you can use the Sentry API to add more data to the event.
> 
> Does this mean that if you just use the `Sentry` nuget, you can do more stuff (add a context and tags) while the `Serilog.Sentry` nuget hasn't exposed this?

Installing `Sentry.Serilog` brings in also `Sentry` so you're free to `using Sentry;` and call directly the Sentry SDK API like `SentrySdk.ConfigureScope(s => s.SetTag("key, "value"));`.

> 
> > If you're on a ASP.NET Core app, and don't install the Sentry.AspNetCore you'll be missing:
> 
> So I like using Serilog because it has the concept of sinks. So (for example) I can get logs going to
> 
> * console
> * seq
> * sentry
> 
> and only have one MS `ILogger` passed around to my code and various reusable nuget libraries that are logging agnostic.

Right, this is how I used [Sentry on NuGet Trends too](https://github.com/dotnet/nuget-trends/search?q=Sentry).

> So this means I would be missing out on these lovely extra's, for an ASP.NET Core app, because I would only be installed `Serilog.Sentry`. Is there any reason why there isn't a `Serilog.Sentry.AspNetCore` nuget?

You can install both `Sentry.Serilog` and `Sentry.AspNetCore`. I did that on NuGet Trends:

https://github.com/dotnet/nuget-trends/blob/d114dc6bb76d729411b6cce45e135ed809257aed/src/NuGetTrends.Api/NuGetTrends.Api.csproj#L18-L19


Here you can see how I initialize Serilog before ASP.NET Core so even if the service fails to start, I'll get an event in Sentry:

https://github.com/dotnet/nuget-trends/blob/4209af5347ad91d1171479df53f624caa634bb79/src/NuGetTrends.Api/Program.cs#L14-L67

Adding both packages also make sure you can can get Sentry through dependency injection.

There's an example of ASP.NET Core and Serilog in this repo, that shows resolving Sentry from the DI container:

https://github.com/getsentry/sentry-dotnet/blob/ba8e98ee2d5c82d1ebf295c38dc310fa941ac9b2/samples/Sentry.Samples.AspNetCore.Serilog/Program.cs#L59-L69

In a real world app you're likely taking the dependency via constructor injection, but the sample exists to give you an idea of what's possible, in a simple way/single file. There's more info about `ISentryClient` and `IHub` [on the docs](https://docs.sentry.io/platforms/dotnet/unit-testing/) but basically adding data to the scope via `IHub` is the same as adding via `SentrySdk` static class, but one will help make your cost testable since you can easily mock `IHub`.

**Comment by PureKrome** [comment]
> Installing Sentry.Serilog brings in also Sentry so you're free to using Sentry; and call directly the Sentry SDK API like SentrySdk.ConfigureScope(s => s.SetTag("key, "value"));

Hmm. okay! Nice! That would work for Sentry-specific logging (and for constant data) while .. the other sinks would not get this info, though? (more on this, below).

Is there any reason why this can't be set TAGS or EXTRA via app settings in a serilog key/value _section_ ? Rational, I try and keep as much configuration stuff _out of the code_ and into the `appsettings.<xxx>.json` files, where appropriate (e.g. constant data, not dynamic data).

which segways now to dynamic data and Logging Scopes :)

does the `Serilog.Sentry` package offer Logging Scopes using the MS default Logger ? Is this what `IHub` is?

for example, all our code _everywhere_ will aways use ctor injection of an `ILogger<T>` and then in our code, [use Logging Scopes like this](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#log-scopes):

(note: I prefer to use key/values with my logging scopes...)

```
public void SomeMethod()
{
    var scopes = new Dictionary<string, string>
    {
        { nameof(transactionId), transactionId },
        { nameof(orderId), orderId }
    };
    using (_logger.BeginScope(scopes))
    {
        // lots of code here.
    } 

    // those scopes are now NOT part of any logging messages, after this point/line.
```
 
> There's an example of ASP.NET Core and Serilog in this repo, that shows resolving Sentry from the DI container:

Okay, so in this example provided, it looks like some middleware that retrieves an `IHub` from DI. Which means ... my code is now _very_ aware that error logging is to Sentry and Sentry only (verus, being logging agnostic). Remember, I prefer to log anything to various sinks (including errors, for example), where Sentry is just one sink in my list. 

So in the above example, if the middleware detects that the current request is trying to go to route "/throw", then do I need to create a Sentry-scope (the `IHub) _and then_ the MS logger scopes, to make sure _all_ sinks capture this? Feels awkward, if that's the case.

In summary, as asked above, is an `IHub` the "sentry way" of doing logging scopes (where tags are indexed and extra's are not) and we can't just leverage the MS Logger 'scopes' instead, keeping our code logging agnostic?

--- 

> Here you can see how I initialize Serilog before ASP.NET Core

Great! this is basically the [same code I have in my SimpleHosting nuget library](https://github.com/PureKrome/SimpleHosting/blob/master/src/WorldDomination.SimpleHosting/Program.cs#L71-L141) so people don't have to repeat that ceremony _every single time they have a new app_. 💪🏻 👍🏻

**Comment by bruno-garcia** [comment]
> > Installing Sentry.Serilog brings in also Sentry so you're free to using Sentry; and call directly the Sentry SDK API like SentrySdk.ConfigureScope(s => s.SetTag("key, "value"));
> 
> Hmm. okay! Nice! That would work for Sentry-specific logging (and for constant data) while .. the other sinks would not get this info, though? (more on this, below).

Yeah if you call `SentrySdk.Something` that won't go into Serilog sinks.

> 
> Is there any reason why this can't be set TAGS or EXTRA via app settings in a serilog key/value _section_ ? Rational, I try and keep as much configuration stuff _out of the code_ and into the `appsettings.<xxx>.json` files, where appropriate (e.g. constant data, not dynamic data).

The ASP.NET Core integration has that:

https://github.com/getsentry/sentry-dotnet/blob/e17703a77a4a5f20b44f233ecfa4e937d42542f8/samples/Sentry.Samples.AspNetCore5.Mvc/appsettings.json#L22-L24

The Serilog integration itself doesn't have `DefaultTags`. We could lift this to `SentryOptions` though, it's common enough. Then you'd be able to bind from your config into the Serilog options.
Do you want to open a PR?


> 
> which segways now to dynamic data and Logging Scopes :)
> 
> does the `Serilog.Sentry` package offer Logging Scopes using the MS default Logger ? Is this what `IHub` is?

Hub contains the scope stack + the Sentry client. So it has `PushScope` and `PopScope`.
The logging integration (MS Logger) does map to it.

> 
> for example, all our code _everywhere_ will aways use ctor injection of an `ILogger<T>` and then in our code, [use Logging Scopes like this](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#log-scopes):
> 
> (note: I prefer to use key/values with my logging scopes...)
> 
> ```
> public void SomeMethod()
> {
>     var scopes = new Dictionary<string, string>
>     {
>         { nameof(transactionId), transactionId },
>         { nameof(orderId), orderId }
>     };
>     using (_logger.BeginScope(scopes))
>     {
>         // lots of code here.
>     } 
> 
>     // those scopes are now NOT part of any logging messages, after this point/line.
> ```
> 

On the latest version of the SDK it should be: 2.1.8. Please let me know if not, it would be a bug.

> > There's an example of ASP.NET Core and Serilog in this repo, that shows resolving Sentry from the DI container:
> 
> Okay, so in this example provided, it looks like some middleware that retrieves an `IHub` from DI. Which means ... my code is now _very_ aware that error logging is to Sentry and Sentry only (verus, being logging agnostic). Remember, I prefer to log anything to various sinks (including errors, for example), where Sentry is just one sink in my list.
> 
> So in the above example, if the middleware detects that the current request is trying to go to route "/throw", then do I need to create a Sentry-scope (the `IHub) _and then_ the MS logger scopes, to make sure _all_ sinks capture this? Feels awkward, if that's the case.
> 

I don't really understand what you said. You don't need to create any scope. That's an optional feature to add some data to a potential event (if something is captured within the scope). When Disposed, the data is also removed (except if an ASP.NET Core request is ongoing btw so we keep it all until the end of the request).

The example is just an example. You don't need to get `IHub` either. It's just to show you how to use the Sentry SDK. If you want to only use ILogger, you can. It's pretty rich in API already with `BeginScope<TState>` and the `Log` method.

> In summary, as asked above, is an `IHub` the "sentry way" of doing logging scopes (where tags are indexed and extra's are not) and we can't just leverage the MS Logger 'scopes' instead, keeping our code logging agnostic?

You don't need to use `IHub`. You can configure how to map the `TState` to Sentry's scope by overriding: `Options.SentryScopeStateProcessor`. Can't do that via json config since it's code.

Here's an example:

https://github.com/getsentry/sentry-dotnet/blob/58670f65317631917cc9fed0ef127abbadc3e108/samples/Sentry.Samples.Console.Customized/Program.cs#L112

ANd:

https://github.com/getsentry/sentry-dotnet/blob/58670f65317631917cc9fed0ef127abbadc3e108/samples/Sentry.Samples.Console.Customized/Program.cs#L272-L287

> 
> > Here you can see how I initialize Serilog before ASP.NET Core
> 
> Great! this is basically the [same code I have in my SimpleHosting nuget library](https://github.com/PureKrome/SimpleHosting/blob/master/src/WorldDomination.SimpleHosting/Program.cs?rgh-link-date=2020-11-29T03%3A49%3A48Z#L71-L141) so people don't have to repeat that ceremony _every single time they have a new app_. 💪🏻 👍🏻

I hope this clarifies things a bit.

**Comment by PureKrome** [comment]
Phew! There's a lot to grok, so lets see if I understand the replies

- `Sentry.AspNetCore` has `appsettngs.json` support for default tags. This is missing from `Serilog.Sinks.Sentry`. A PR to accept this would be considered.
- MS Logger supports `BeginScope` and Sentry (2.1.8+) respects this. So if we use MS Logger + `BeginScope` + Sentry, we should see the scoped data in Sentry.
- Because Sentry respect the MS Logger scopes, by inheritance, the `Serilog.Sinks.Sentry` nuget library _should also_ handle MS Logger + Scopes.  

Does that sounds about right?

**Comment by bruno-garcia** [comment]
> Sentry.AspNetCore has appsettngs.json support for default tags. This is missing from Serilog.Sinks.Sentry. A PR to accept this would be considered.

We could move default tags to `SentryOptions` so all integrations could leverage.

> MS Logger supports BeginScope and Sentry (2.1.8+) respects this. So if we use MS Logger + BeginScope + Sentry, we should see the scoped data in Sentry.

It always had it, but before 2.1.8, on ASP.NET Core (with the scope locked) it didn't keep the values. See #594

> Because Sentry respect the MS Logger scopes, by inheritance, the Serilog.Sinks.Sentry nuget library should also handle MS Logger + Scopes.

Serilog does some funny stuff with ASP.NET Core with regards to logging backend. I see the Sink itself adds some Serilog properties stuff as Extra: https://github.com/getsentry/sentry-dotnet/blob/e17703a77a4a5f20b44f233ecfa4e937d42542f8/src/Sentry.Serilog/SentrySink.cs#L109

I suggest trying stuff out though.

**Comment by PureKrome** [comment]
Great - created an issue (in preparation for a PR) to test this stuff out, etc.
